### DISCUSSION TITLE
On Dictionary Search

### DISCUSSION THREAD

**Discussion by burnsauce** [jkotlinski/durexforth#428] [General]
Searching the dictionary will inevitably be a number of string comparisons.   We can control 2 things about this process, performance-wise:
1. How fast each string comparison is
2. How many string comparisons are made

Currently, our string comparison function is relatively fast.  It filters first by length, then character-by-character until a branch on mis-match, or a fall-through match.  Not much could be done here to improve the speed of string matching, if anything.

We have many comparisons to make, however.  If `wdepth` is the word's distance from `latest` in the dictionary, then we have to make `wdepth` comparisons to find that word.  New words lookup quickly, while older words lookup slowly.

This has bad effects for common, builtin words, as the most fundamental words take the longest to find.  Complicating the issue is the Forth Standard, which specifies that newer entries need to be found before later entries in the case of duplicate words.

It's a wrench in the gears of the common response to string searching: sorting the word list in some fashion to reduce the number of comparisons made.  If we do not retain inverse word definition order when calling `FIND`, Forth doesn't work correctly.

So, if we are to employ sorting at all, we must include data about the definition sequence in the `COMPARE` operation of whichever algorithm we might seek to employ, growing the dictionary size.

Here we see illustrated the trade-off we face with dictionary search:  speed for size.  There may be a much faster dictionary structure, but it **will** be larger.

**Comment by jkotlinski** [comment]
"It's a wrench in the gears of the common response to string searching: sorting the word list in some fashion to reduce the number of comparisons made. If we do not retain inverse word definition order when calling FIND, Forth doesn't work correctly."

Actually, that does not need to be a problem. If a stable sort algorithm is used, like merge sort, then words with the same name could keep their relative order.
The bigger problem is how to preserve functionality of words like `marker`.

**Reply by burnsauce** [reply]
Yep, that's one way to unwrench it.

`marker` is fine if dictionary entries are sequentially numbered.  Same size->speed tradeoff.

**Comment by jkotlinski** [comment]
Similarly, I considered to make search faster by storing a 16-bit string hash instead of the full string.
That would work fine, except for `words`.

**Reply by burnsauce** [reply]
As long as the hash was fast and collision-free. Would you stop storing the actual string?  `see` no evil...

**Reply by burnsauce** [reply]
Using this algorithm:
```
int hash = 0;
for (int i = 0; i < s.length(); i++)
    hash = (R * hash + s[i]) % M;
```
I am searching for a set of R, M primes that fit the dictionary after `include test` with no `--test--`, a dictionary of over 700 words.

In a live durexForth environment, it might take me a while.

Let me know if you have any insight.

**Reply by Whammo** [reply]
Could a CRC serve the same purpose?

**Reply by burnsauce** [reply]
If the CRC applied to every word in the dictionary is unique, yes!

**Reply by Whammo** [reply]
Would it be calculating over the string, the first letter, the whole word? 
 Maybe just the length byte and the string.

**Comment by burnsauce** [comment]
Not yet mentioned is the effect of this sorting on item 1: string comparison.

The `COMPARE` function of the search always needs to do more than comparing the string length.  It still has to be efficient to outperform the current design, even if it is O(log n).

**Comment by Whammo** [comment]


**Reply by burnsauce** [reply]
Care to edit this post?  I don't understand what you mean.  Each word already has a unique address.

**Reply by burnsauce** [reply]
Pre-compiling the durexforth source as a cross-compilation process would indeed speed up build times, but do nothing for compilation of source at runtime.

**Reply by burnsauce** [reply]
And yeah it's a little off-topic.  The subject is searching the dictionary at runtime, sorry if that wasn't clear :)

**Comment by Whammo** [comment]


**Reply by burnsauce** [reply]
If the addresses are so "fixed" in source, how would you handle different include orders?

**Reply by burnsauce** [reply]
Imagine this simple example:

A user writes some code in `v` and compiles it, then saves his source file to disk.

He later goes back and modifies the source code to add a word up at the top.

What becomes of all the pre-compiled source addresses?

**Reply by burnsauce** [reply]
Wow, sounds like quite a design.  Let me know when you have it working, I would be interested to see.

**Reply by Whammo** [reply]
I see now that the context of this thread was set plainly in the beginning of this thread and in that context my comments WERE intrusive and off-topic. I will try to do better in the future and humbly apprentice.

**Comment by jkotlinski** [comment]
The design mentioned by whammo (editor does compilation instantly) actually
is not new. Maybe colorforth does something similar?
Im sure OKAD does

**Reply by Whammo** [reply]
I believe colorforth does, but perhaps it's like its storing the dictionary in arrays. To do it efficiently, requires higher-order instructions the 6502 doesn't have.

**Comment by Whammo** [comment]
One way to reduce comparisons when searching for a word is by first comparing the length.
Then you compare the first letter. 

EDIT: That's why FIND-NAME does that.

**Comment by lonetech** [comment]
There's an order-preserving hash table in [CPython 3.6's dict type](https://docs.python.org/3.6/whatsnew/3.6.html#whatsnew36-compactdict). The essence of it is fairly obvious, separating the hash-indexed array from the key array. This index could grant hash speed lookups into a dictionary, and you could fall back to linear search if the hash index didn't fit. E.g. we could build the hash table by traversing the words and not overwriting hash entries that are found. Then, if the lookup by hash finds a mismatched word, we can search from that position anyway as the word we want is either missing or deeper. Alternatively, we could traverse misses from latest as now, and update the hash table for a most recently used symbol cache. Best of all, these methods could be patched on and transient. 

Something like:

    : hash ( caddr u -- c ) ( some compact hash function) nip ;
    variable hashtable
     \ Could be reorganized into ( -- hashtable | 0 ) for use with ?dup if
    : ?hash-valid ( -- flag ) 
      hashtable @ dup
      here u> swap
      $200 + latest =
    and ; (allow hashtable to be discarded when dictionary changes)
    : find-name-h ( caddr u -- 0 | nt )
      ?hash-valid if
        2dup hash words hashtable + @  \ lookup hashed entry
        ( Most recent used cache: check entry, full find if wrong, update.
          Most recent defined: do find from entry, return 0 if not found. )
        ?dup if
          latest >r  to latest  find-name  <r to latest   \ hack: let find-name do verification, and deeper search
        else
          exit  \ hashtable already confirmed not found. Alternative: full find-name. 
        then
      else find-name then ;
    : update-hash ( nt -- flag )
      dup c@ 1+ swap hash 2*  ( nt hash )
      hashtable @ + ( nt hashentry )
      dup @ if 
        2drop  \ guard against overwrites
      else
        !
      then
      -1
    ;
    : rehash ( -- )
      latest $200 - 
      dup here < if drop exit then  \ guard against out of memory
      hashtable !
      hashtable $200 0 fill
      ['] update-hash dowords
    ;

It's just a rough outline. Not tested. dowords will be helpful in building the hash table. The suggested latest check is vulnerable to specific forget/marker/define combinations that land on the same length again. Workaround: always rehash after forget or marker. Alternative for static hashtable: update-hash along with define. Actually, invalidate the whole table on any forget, because it changes the nametokens for everything newer. 

Another method to reduce comparisons is not to search the full dictionary. Forget does this by pruning internal words, and [search order word lists](https://forth-standard.org/standard/search) can do so for larger sets. I think that extension is reasonably implementable. Our current find (or find-name) stops when it encounters a 0. If we add the pointer back in to that 0-length entry, we can use it to indicate the next word list to search. That's 2 bytes per word-list, and a small extra check making the difference between find and search-wordlist. Set-order can then rearrange these word-list links. The wordlist identifiers can simply hold the beginning of each dictionary. The current compilation wordlist can be the latest, or we could make set-current actually move the word-list to the lowest address. latest will have different values for define and find.

**Reply by lonetech** [reply]
FWIW, now I have tested some code. It appears it takes some inline 8 bit code to do, but an average speedup looks plausible:

    \ Accelerate name lookups with hash table

    \ Need a hash function that is fast and preferably compact.
    \ Pearson hashes are reasonably strong, but require minimum 4 cycles per byte for lookup.

    : nt>au dup 1+ swap c@ ;

    code hash ( caddr u -- c )
    \ idea: asl and sbc. asl will set carry flag, sbc folds it back in.

    \ Setup: load address in w3, length into y
    lsb ldy,x
    lsb 1+ lda,x
    w3 sta,
    msb 1+ lda,x
    w3 1+ sta,

    dey,    \ for 0-index on caddr. Not needed for nt / cstr!
    \ Now we have caddr in w3 and u in y. loop
    0 lda,#
    :-
    asl,a       \ 2 cycles
    w3 sbc,(y)  \ 5+ cycles!
    dey,
    -branch bpl,

    \ teardown: put result on stack
    inx,        \ drop
    lsb sta,x
    0 lda,#
    msb sta,x
    ;code

    \ TODO: place hashtable. e.g. latest $ff00 and $200 -
    \ Aligning to $100 avoids page boundary penalty
    $200 allot
    here $ff00 and $100 -
    constant hashtable

    : nt>h? ( nt -- -1 )
    \ Update hash entry if not already filled
    dup nt>au hash  \ calculate hash for nt
    $fe and    \ single page hashtable (128 entries)
    hashtable + dup @ ( nt he hev )
    if \ entry present
      2drop
    else
      !
    then -1 ;

    : rehash
    hashtable $100 0 fill
    ['] nt>h? dowords ;

    : find-name-h ( caddr u - 0 | nt )
    2dup hash
    $fe and    \ single page hashtable (128 entries)
    hashtable + @
    ( caddr u hev )
    ?dup if
        latest >r
        to latest
        find-name
        r> to latest
    else
        2drop 0
    then ;

    code find-name-hc ( caddr u - 0 | nt )
    \ Setup: load address in w3, length into y
    lsb ldy,x
    lsb 1+ lda,x
    w3 sta,
    msb 1+ lda,x
    w3 1+ sta,

    dey,    \ for 0-index on caddr. Not needed for nt / cstr!
    \ Now we have caddr in w3 and u in y. loop
    0 lda,#
    :-
    asl,a       \ 2 cycles
    w3 sbc,(y)  \ 5+ cycles!
    dey,
    -branch bpl,

    \ Lookup entry
    $fe and,#
    tay,
    hashtable lda,y
    w3 sta,
    hashtable 1+ lda,y
    w3 1+ sta,
    \ Possible optimization: entries are not in zero page,
    \ so zero test can be done on MSB only
    w3 ora,

    +branch bne,

    \ Entry empty, so not found
    \ ' find-name jmp,
    inx,
    0 lda,#
    lsb sta,x
    msb sta,x
    rts,

    :+
    \ Have an entry, search from there
    \ find-name will check full name
    \ Save latest
    ' latest 3 + lda,
    pha,
    ' latest 1+ lda,
    pha,
    \ Start search from cached entry
    w3 lda,
    ' latest 1+ sta,
    w3 1+ lda,
    ' latest 3 + sta,
    ' find-name jsr,
    pla,
    ' latest 1+ sta,
    pla,
    ' latest 3 + sta,
    ;code

Demonstration code (uses timer.fs):

    : fnt0 nt>au find-name ;
    : fnt1 nt>au find-name-h ;
    : fnt2 nt>au find-name-hc ;

    : verify nt>au
    2dup find-name >r
    2dup find-name-h >r
    2dup find-name-hc
    dup r> = swap
    r> = and
    if
      2drop
      -1
    else
      ." !"
      type
      0
    then ;

    require timer

    : hashdemo
      rehash
      60
      ." find-name   " start over 0 do
      ['] fnt0 dowords loop stop cr
      ." find-name-h " start over 0 do
      ['] fnt1 dowords loop stop cr
      ." find-name-hc" start over 0 do
      ['] fnt2 dowords loop stop cr
      ." verify" ['] verify dowords cr
      drop
    ;

Example results:

    hashdemo
    find-name   7.566
    find-name-h 8.633
    find-name-hc4.783
    verify
    ok

**Reply by Whammo** [reply]
This looks like something that could find it's way into the 4k RAM under I/O.

**Reply by lonetech** [reply]
With a luxurious 4k, I'd split the MSB and LSB (like the data stack) and use the full 8 bit hash, removing the `$fe and,#`. Twice as many entry points and two cycles faster. Anyhow, from here the next step is patching calls to load-name, and adding an update call to `header`, `;` and `:code`, so the hash table can be kept current during `included`. Another improvement would be self modifying the pointer instead of indirecting through W3, to save a cycle per character in the hashing routine.

**Reply by lonetech** [reply]
I just had another silly little thought. The find function could be sped up if the earliest comparison it makes held more information. If I hash all but the first character of a word, and mix that hash in with the first byte, it would reduce the rate of false positives because there would be 256 possible values in the byte rather than about 68. The downside is that name tokens would require additional decoding for display, but I think that's only in `words`, seldom used; we could just add a nt>s decoder word. Oh, and this is very much a weird train of thought; I was thinking of how I could improve dictionary search performance without making it larger. 

The base idea was to have words checked for matching hash, not just length, before going full comparison. 

Come to think of it, this process is inductive. We could make every character mix in every later character. And that's even easy to build, as you'd do a scan from right rather than a fold. And the hash function already processes from the right to simplify the loop condition.

**Comment by Whammo** [comment]
How would we search and maintain the directory if we had unlimited memory, and we want compiling as fast as possible?

**Reply by lonetech** [reply]
With unlimited memory, we'd simply allocate a huge region of memory for each possible word. Just convert the word straight into an address. Markers can be allowed by simply storing a copy of the dictionary (copy on write virtual memory may be of use). 

If we e.g. assume words are made up of printable ASCII, that's 95 symbols, so with a length up to 31 characters we'd be looking at about log(95**31,2)=204 bit indices into the dictionary array. IIRC there are Forths that use only three characters from the name, which is a more manageable 20 bits but obviously a pretty severe restriction. 31 is taken from the standard and matches durexForth (`STRLEN_MASK` in durexforth.asm). The standard is case insensitive, which reduces the potential character set by another 26, but the difference isn't super significant (durexForth implements that by lowercasing words). 

This is effectively the extreme case of a hash table; one where the hash is a 1 to 1 representation of the name itself, and the table large enough to hold every possible hash value. No collisions can occur, so no collision resolution strategy is needed.

**Reply by Whammo** [reply]
Thank you for your thoughtful answers. The actual problem doesn't seem so large now.

**Comment by jkotlinski** [comment]
Just to loop back into the conversation. Hashing is a fun idea, but realistically I do not think it will give any dramatic improvement. This is because what we already have, string length + first character, is already a fairly decent 16-bit hash.

Implementing the Search-Order word set, like Whammo suggested, should give a very good improvement. This is obvious, since the assembler words alone take up more than half of the word list. Switching them out should give a notable speed-up.

**Reply by jkotlinski** [reply]
Created issue #476

**Reply by jkotlinski** [reply]
Of course, implementing the Search-Order word set will take some RAM. So what is won in speed, might be lost in size. But maybe it is not too bad?!

**Reply by lonetech** [reply]
It's 16 bits of memory, but the hash strength is closer to 5+6=11 bits. Good point though. Also, the hash *table* is a different story, enabling on average skipping half the dictionary each search. As you say, that's a smaller improvement than excluding more than half the entries to begin with. 

Our dictionary structure isn't the linked list in Starting Forth, though. We have a packed list with length fields for traversal. Furthermore, the length field has other bits reused, so we can't just set a big length on one word to hide a covered set. 

What we can do is decide no collisions exist in the shipped words, and therefore that their order doesn't matter. With that in mind, we could hide words by moving them *after* the dictionary, as long as nobody is keeping name tokens around across such changes. Another thing we can do is place words elsewhere than core. 

So, for a module we require in building but may never use at runtime like asm:
1. Measure size by compiling once and comparing here and latest before and after.
2. Prepare space for a relocated version by relocating dictionary down.
3. Set here to the space above the dictionary, and compile again.
4. Reorder the dictionary to have the asm words at the end, near the code. 
5. Restore here to its original value. 
6. Hide those top words by writing the 0 byte that causes find-name to stop searches. 
7. Reveal the top words by restoring the byte replaced with 0. 

This way, the module can even be stripped by the turnkey utilities, as it is nothing but dead space above the dictionary while hidden. 

Of course, the reordering and hiding are easier steps in this, and this isn't the full Search-Order word set.

**Reply by jkotlinski** [reply]
Oh yes - a hash table is a different thing... Sorry for not reading the entire conversation carefully.

The problem with hash tables is, I guess, that standard Forth more or less expects a linked list. To allow for multiple definitions of the same word, and that order matters.

**Reply by lonetech** [reply]
That doesn't really concern the hash table as I implemented it, which is just an index of where to start looking. I just fall back to the old search routine from the latest word to have the correct hash.

**Reply by jkotlinski** [reply]
OK, I see now! Sorry for being so dense. Yes, that is smart!

**Reply by lonetech** [reply]
It's not super smart, but proof of concept stuff. What I found was that this would on average speed up references to defined words. It still leaves a lot of dictionary traversal, and scales badly; if you just define a word for each hash value, you'll still always be traversing the entire dictionary below those.

**Reply by lonetech** [reply]
Note: One possible optimization is to build a hash table for whatever suffix of the word list contains no collisions. That would allow a single check fall back, and mix nicely with @burnsauce's idea to section off suffixes. TODO: Write tools to build such suffix hash tables and report how many words it fits per table. Tables may overlap in found words; the sooner a hit is found the better.

**Reply by lonetech** [reply]
A quick check with my current hash function shows at least three 6-fold collisions, yet lots of slots have nothing in them. The hash function is simply not great, expected as it was designed for speed and size. This means a smarter resolution than duplicate tables might be called for, but for starters, 6 tables takes 3KiB, so would fit in the $d000 hidden RAM.

**Reply by Whammo** [reply]
4k of memory at $d000 to trade for speed. Are there loops to unroll?

**Reply by lonetech** [reply]
Why yes. There's a hashing function for which 6 characters would suffice (along with length), and unrolling it lets us do distinct operations for each character, as well as leave information in the carry flag - which means we could maybe fit all 340 base words in one collision free 1KiB hash table (ignoring the actual duplicate words like ;code and jmp, from labels). Also the word to name conversion in header and find-name could be factored out and inline the lowercase routine.

**Comment by lonetech** [comment]
Here's a silly little proof of concept that can hide the assembly words:

    : lift-words
      parse-name find-name
      dup c@ $1f and + 3 +   \ older
    
      parse-name find-name   \ newest
    
      2dup -                 \ size
      dup latest swap -      \ tmp
    
      2 pick   \ newest
      1 pick   \ tmp
      3 pick   \ size
      move
    
      latest                 \ latest
      4 pick latest -        \ bothsize
      move
    
      drop drop drop
    ;
    
    \ Rearrange the dictionary
    lift-words adc,# -branch
    lift-words pushya does>
    lift-words -rot lift-words
    
    : asm
     [ parse-name -branch find-name
       dup literal c@ literal ]
     over c@ xor swap c! ;

The `asm` word here toggles the `-branch` entry between end of dictionary and the start of assembly words. It is precisely the sort of code that makes this dictionary rearranging trick risky, because it keeps a pointer to where `-branch` was when `asm` was compiled. 

The reason this has a lift-words routine rather than a bury-words is to perform the entire dictionary change in one move. I'm too tired to figure out the proper sequence of it all now.

**Comment by lonetech** [comment]
FWIW, I think I now have some evidence that `find-name` is where we spend our time. Based on an execution trace of the base compilation:

```
>>> where.groupby('name').sum().sort_values('cycles').tail(50)
                      cycles
name                        
word                 16632.0
.compile:comma:      26773.0
drop                 29553.0
.getc                36536.0
.here                38360.0
.:lparen:            38598.0
.key                 39953.0
.:dot::lparen:       42317.0
:gt:                 58582.0
tuck                 59230.0
:equals:             65398.0
.included            65828.0
.:gt:in              99530.0
.header             103417.0
quit                123208.0
.litc               123950.0
.c:comma:           139505.0
min                 145269.0
.:comma:            167181.0
:two:dup            170834.0
:lt:                198124.0
:one:               222747.0
ud:slash:mod        241669.0
:store:             259181.0
um:star:            334966.0
:fetch:             377678.0
.source             418296.0
interpret           426365.0
pushya              477828.0
.branch             514313.0
:zero::equals:      548746.0
parse:minus:name    599343.0
execute             603210.0
.ioabort            609081.0
.lit                679459.0
c:fetch:            846151.0
r:fetch:            925181.0
:plus:              935581.0
.:slash:string      998629.0
:minus:            1078964.0
dup                1199980.0
over               1480184.0
:gt:r              1640455.0
r:gt:              1709267.0
.:zero:branch      1718133.0
.saveb             2191946.0
swap               2297198.0
:gt:xt             2795020.0
.refill            3739136.0
find:minus:name   28892893.0
```

I expect `refill` is reading the source code, `saveb` is saving our final program, and `>xt` is called by `interpret` if `find-name` found a word. The leading periods in the names are artifacts of me parsing the label file slightly wrong. Also, my profiler is rather unpolished, with a 1.9G monitor log to parse. Overall `find-name` was responsible for 47.5% of the execution time.

**Reply by jkotlinski** [reply]
That seems probably right. I imagine this loop is very hot:

```
.examine_word
    and #STRLEN_MASK
.findlen
    cmp #$ff ; overwritten
    beq .string_compare
.string_compare_failed
    ; no match, advance the dp
    clc
    adc #3
    adc W
    sta W
    bcc +
    inc W + 1
+   lda (W), y
    ; Is word null? If not, examine it.
    bne .examine_word
```

One idea I got now - the `and #STRLEN_MASK` could be avoided, by moving the `F_IMMEDIATE` and `F_NO_TAIL_CALL_ELIMINATION` flags from the length byte to the last character of the word. That would save two cycles per length byte comparison.

Another observation: `lda (W),y` has a one-cycle penalty when it crosses pages. It could be investigated, if it is more efficient to increment y and keep W zero... I doubt it is an overall win though.

**Reply by lonetech** [reply]
That's precisely the hottest part, at 41.3%. `.string_compare` appears to get called in 11% of iterations. 

I'm not sure we have two bits to spare in a name byte, though; I estimate the available character set to be 138 or 69 characters (by case folding and ignoring control characters and spaces). I do agree that isolating length can be valuable; it's always needed for the iteration even on misses. 

Incidentally, the string compare itself could be accelerated by REU verify mode, at the cost of writing three bytes and reading one in IO space. 

The one cycle carry penalty I think isn't worth worrying very much about. It'll only apply during checks of names that cross page boundaries. The other way around would involve checking for carry after incrementing y. If we're hunting per-character costs in the string compare, `lda (W), y` is 5+ cycles while absolute,y mode is 4+, so copying W to an immediate operand would consistently save one cycle per iteration there. The problem is, that copy itself probably costs `2*3+2*4=14` cycles, and most words aren't that long. 

One moment. We're checking for equality. Why are we using arithmetic CMP instructions? An EOR would produce Z flags just as easily without requiring carry setup. Since the dictionary doesn't wrap around to zero page, we never have carry set after increasing W. EOR and DEY don't touch carry. We could remove CLC from the loop!

Not a completed thought though, the length is after all used as input to the addition immediately after. One more EOR would restore it, and we'd replace TAY with LDY in .string_compare, leaving 3 places to store the length instead of 1, and just two cycles saved from moving .string_compare_failed down one instruction. 

The real gains are in reducing the number of words to check.

**Reply by jkotlinski** [reply]
Looking at a PETSCII table, it seems all durexForth word names use characters in $21-$5F range. With that restriction, we have two spare bits.

EOR sounds like a great idea!

Another optimization idea: the `adc #3` can be removed by storing the +3 in the length byte itself. Meaning, '!' would have length byte 4.

**Reply by lonetech** [reply]
So you mean restricting the word character set to screen codes <$40? That would simplify the lowercase code immensely. 

Folding in the constant 3 is good. That would offset the cost of destructive eor compare on the length. 

And if we have 2 bits free in name characters, we can fold in the last characters of long strings, sort of reverse base64 style. That would crunch down a 31 character name to 24 bytes. Algorithmically it wouldn't be super complicated; each time len>=4, take last character, store its bits in the first three, advance base pointer and reduce length. Which gives us the space saving formula: one byte per 4 characters, which includes all assembler mnemonics, and discriminates addressing modes faster, as well as start/state, again/allot, chkout/clrchn/create, etc. 

Only 96/348 words are shorter than 4 characters. This hack would save 274 bytes in the default dictionary and 22 bytes in CHAR_TO_LOWERCASE, but would of course cost a bit in FIND_BUFFER postprocessing. Which would also need to be done in HEADER. And name>string would need to decode it. 

CHAR_TO_LOWERCASE has a third use in READ_NUMBER, which may actually want to be pickier about unexpected characters. If not, you'd basically modify it to compensate for the different offset between 0 and a (-47 instead of 17). 

A complication of reducing stored size is that stored bytes no longer matches with name length. This might be possible to compensate for by exploiting that space is in the character set; a 3-byte name could be padded to 4 bytes with a space, before folding it in. Otherwise you'd have `foo` and `foo@` encode to the same name (or something more complicated, as the original bits will probably be shifted due to lack of registers in 6502).

**Reply by jkotlinski** [reply]
Yes exactly, all word name characters could be stored as PETSCII $20-$5f (I suppose these map indirectly to screen code 0-$3F). Upper-case is still allowed, but must be discarded as is now, as Forth is case insensitive.

If the goal is to speed up the full durexForth build, probably the most bang for the buck would be to convert the Forth source code to assembly.

**Reply by lonetech** [reply]
They map through this function, for that subset of the characters: `screencode = petscii&0xc0 ? petscii&0x1f : petscii;`

Having recently browsed the [finding definition names](https://forth-standard.org/standard/usage#usage:find) section, it's implementation-defined if Forth is case sensitive or not. However, the standard names are described in upper case, and must be found if written as such. The C64 PETSCII dualcase code page complicates that a touch by switching what characters are upper or lower case, ending up with one lowercase and two uppercase sets in the mode durexForth uses. That in turn complicated the lowercase function, which maps all three alphabets onto the $41 range. But it's simpler to map them to screencode lowercase, where they start at 1, and as a bonus we get the two top bits clear. 

I'm not greatly concerned with the speed to build durexForth itself; that takes three seconds when I'm not profiling. I simply find the challenge of optimizing it to be interesting. 

```
$ time make clean deploy
rm -f *.lbl *.prg *.d64
rm -rf build deploy
mkdir -p build
touch build/empty
echo  >build/c1541.script format "durexforth,DF" d64 durexforth.d64
...
Output file: deploy/durexforth-v3-0-0-M.crt
Conversion from binary format to C64 Simons' BASIC .crt successful.

real	0m3.281s
user	0m2.979s
sys	0m0.160s
```

**Reply by jkotlinski** [reply]
Right, the "make" part is decently quick. What takes more time for me is compiling Forth code on C64. Not that it is so terribly bad, either.

I have not been that eager to do optimizations lately. By definition, optimizing means improving something by making something else worse. For durexForth, it could mean making it faster by adding complicated code or consuming more RAM.

For the idea about moving word flags from length byte to name tail, I feel that it is _possibly_ not an optimization, but just making it plain better. I guess only way to find out is to try to make the change, and see how it turns out.

**Reply by burnsauce** [reply]
> For the idea about moving word flags from length byte to name tail, I feel that it is _possibly_ not an optimization, but just making it plain better. I guess only way to find out is to try to make the change, and see how it turns out.

I'm excited to see how it would look!  Would the length field be a full 8 bits now?

e: Rereading the discussion, I'm more compelled to say this:

If we're crunching data in the dictionary and accepting the corresponding en/decoding tax on string data, I wonder if stuffing other data in there instead might work better.

If we populate a search tree, users could load an accelerator at runtime.  Maybe _it_ could sit under I/O?

**Reply by lonetech** [reply]
This was `make deploy` - including running Vice headless to do the base compilation before converting to cartridge, as discussed in #312 and implemented in my [faster-build branch](https://github.com/lonetech/durexforth/tree/faster-build). The Vice x64 execution which compiled `base` took less than a second. The host is a first generation GPD Win Max running Debian GNU/Linux. 

The faster-build approach does have an issue with creating a definition just to hold an invalid instruction to terminate running after turnkey finishes. I think that could be improved using return oriented programming, which looks remarkably sane in Forth. Particularly if we use the [debugcart](https://vice-emu.pokefinder.org/wiki/Debugcart) instead of JAM. 

I disagree with your definition of optimization. You describe trade-offs, such as loop unrolling trading execution speed for program size, but optimizing itself is working towards an optimum, per my Webster's. E.g. if we find code that performs a task in a needlessly complicated manner, the cost is the work of finding the better solution, which could improve the program by any metric. We're only looking at tradeoffs if we optimize for one metric over another. 

I believe switching dictionary encoding from PETSCII to screen codes is a net win for the lowercase routine in particular, though there is a cost in `name>string` used by `words` and `dump-labels`. Point being, every word lookup *including* `words` becomes cheaper, as well as the `header` for declaring every word. So much cheaper that you'd have to rerun `words` hundreds of times for the tradeoff to manifest. It's replacing code with smaller, less often run code. 

I'm less certain about the 6 to 8 bit packing, which carries a new execution cost while converting names for lookup, but it would be faster on longer names and could be implemented in the style of Duff's device. 

The key property is that we're making dictionary searches for every word, and they already use the lowercase routine which is costly as is. It's relaxing the requirements on it to use the 63-character set we've discussed that saves such a large portion other complications become plausible. 

Either way, with the profiling tool set up (it's still clumsy at about 90 seconds runtime) we should be able to produce empirical measurements. 

@burnsauce What sort of accelerator did you have in mind? Bear in mind executing on a stock C64 is a rather core feature. Also, building and balancing search trees is quite the task. 

I did design the hash table lookup as a software approach to such optional acceleration, but it doesn't scale great and should be hooked in after lowercase, not before.

**Reply by burnsauce** [reply]
> @burnsauce What sort of accelerator did you have in mind? Bear in mind executing on a stock C64 is a rather core feature. Also, building and balancing search trees is quite the task.

Building and balancing is work, yes, but searching is not.  If the "stock" words have their tree created at build time, there's no impact to the user.  If the accelerator were simply a redefinition of `find-name` to search this tree before falling back to a linear dictionary search that terminates at the core words, and no updates to the tree were possible at "runtime", this would be almost all upside.

... assuming that such a tree would fit in the stock dictionary and that searching it (including decoding) was faster than `find-name` currently is for _core_ words.

**Reply by lonetech** [reply]
Ah, I understand! And this would be possible without insane effort, too. We could keep the old find-name for the user dictionary, just put an end marker between it and the core dictionary, and call the core search-tree function if find-name failed to find a user word. (Forth ordering is that latest words take priority.) The same approach works for the hash system too, and as bonus we get the lowercase name prepared in the find buffer. And this strategy only requires one injection at `.find_failed` in `find-name`, where there's already a jmp.

**Reply by burnsauce** [reply]
> (Forth ordering is that latest words take priority.)

Ah yes of course, how could I forget?

**Comment by lonetech** [comment]
Proof of concept verification for simpler `lowercase` routine:

```
code lc-core
  \ Asm core routine size: 29 bytes
  \ Shortest path: 7 instructions
  lsb lda,x

  $15f6 jsr,

  lsb sta,x
;code

code lc-prop
  $c0 lda,#    \ This setup let us use BIT.
  w   sta,
  lsb lda,x

  w       bit,
  +branch beq,
  $1f     and,#
  $40     ora,#     \ Remove this instruction for $00-$3f screen codes
:+

  lsb sta,x
;code

: check
  dup dup
  emit lc-core emit lc-prop emit
;

: verify
  2dup
  cr ." Verifying range " . .
  do
    i lc-core
    i lc-prop
    <> if
      cr i . i check
    then
  loop
;

cr $60 $20 verify
cr $80 $60 verify
cr $e0 $c0 verify
```

By restricting the problem domain to the symbols we discussed, this reduced the lowercase routine core to four instructions. A few PETSCII graphics code points were converted into arrows, brackets, fetch and comment characters. As a bonus, this converts the ASCII lowercase range also. The main decision is where to store the $c0 constant for BIT; it would be more efficient to have a permanent value in zeropage, but anywhere could do. Not using BIT means we'd need to save the original value, which the current CHAR_TO_LOWERCASE does. Then again, putting it out in code memory and still using BIT just costs two bytes and one cycle per call, far faster than any save/restore.

**Reply by lonetech** [reply]
Implemented a simplified lowercase routine optimized for the used/permitted character set. 

hits time
16470490 60798409
hits time
16345082 60458637

That's a 0.5% speed improvement in base compilation for using the shorter lowercase routine. It reported 901 bytes remain (up from 883) and still recognizes mixed case word input. The downside is it will happily swallow a variety of control codes, graphics symbols etc and turn them into less invalid names. E.g. non-breaking space now turns into @. Please don't use that in source code. 

https://github.com/lonetech/durexforth/commit/e0608d680415cee850b239a93bbc4e929ef1cc9f
